RELEASE/scripts/autoscend/auto_quest_level_12.ash

```diff
@@ -1290,6 +1299,223 @@ boolean L12_themtharHills()
+		if(faction == "fratboy")
+		{
+			enemiesRemaining = 500-get_property("hippiesDefeated").to_int();
+		}
+		adv_saved = ceil(enemiesRemaining.to_float() / (base_kills + bonus_kills)) - ceil(enemiesRemaining.to_float() / (2*base_kills + bonus_kills));
+	}
+	
+	//is it worth it to farm a chaos butterfly?
+	if(!get_property("chaosButterflyThrown").to_boolean() && item_amount($item[chaos butterfly]) == 0)
+	{
+		//softcore pull?
+		if(!in_hardcore() && adv_saved > adv_needed-15)
+		{
+			if(pullXWhenHaveY($item[chaos butterfly], 1, 0))
+			{
+				return true;
+			}
+		}
+		
+		//4 enemies in [The Castle in the Clouds in the Sky (Ground Floor)] mean there is a 25% chance to encounter the one we want.
+		//roughly estimate 4 turns per possibility giant encounter. at base drop this means ~20 adv needed.
+		if(is100FamiliarRun())
+		{
+			addToMaximize("200 item drop");
+		}
+		else
+		{
+			addToMaximize("200 item drop, switch Baby Gravy Fairy, switch Jumpsuited Hound Dog");
+		}
+		float expectedItemDropMulti = 1 + simValue("Item Drop")/100;
+		int adv_to_get_butterfly = ceil(20 / expectedItemDropMulti);
+		if(adv_saved > (adv_needed + adv_to_get_butterfly))
+		{
+			if(autoAdv(1, $location[The Castle in the Clouds in the Sky (Ground Floor)]))
+			{
+				return true;
+			}
+			else
+			{
+				auto_log_warning("For some reason failed to adventure in [The Castle in the Clouds in the Sky (Ground Floor)] for a [chaos butterfly]... skipping", "red");
+			}
+		}
+	}
+	
+	//if you just now acquired [chaos butterfly] and don't have anywhere to use it other than the barn or the war then use it in barn
+	//also contain code to account for being unable to use it for some reason
+	if(!get_property("chaosButterflyThrown").to_boolean() && item_amount($item[chaos butterfly]) > 0 && auto_my_path() != "Pocket Familiars")
+	{
+		if($location[McMillicancuddy's Barn].turns_spent > 0)
+		{
+			auto_log_warning("I seem to have spent turns in [McMillicancuddy's Barn] without using the [chaos butterfly] which I have. Something is not right...", "red");
+		}
+		else if(adv_saved > adv_needed-15)
+		{
+			auto_log_warning("Looks like I should use the [chaos butterfly] in [McMillicancuddy's Barn]", "blue");
+			if(autoAdv(1, $location[McMillicancuddy's Barn]))
+			{
+				return true;
+			}
+		}		
+	}
+	
+	//final check and then we can start doing the farm
+	if(adv_saved > adv_needed)
+	{
+		auto_log_info("Save McMillicancuddy's Farm from the Dooks", "blue");
+		set_property("choiceAdventure147", "3");	//open the pond
+		set_property("choiceAdventure148", "1");	//open the back 40
+		set_property("choiceAdventure149", "2");	//open the other back 40
+		
+		//because of free fights we can't rely on adventures spent to see if a zone is clear. so try each zone in turn. 
+		//if it successfully adventures there then it will return true and restart the loop. 
+		//If it fails all zones then assume quest is complete and turn it in
+		if(autoAdv(1, $location[McMillicancuddy's Barn]))
```

Comment:
This is super awkward. We could maintain an internal property `auto_duck_stage` which starts at `0` and is advanced to `1` when we can no longer adventure in the Barn, etc. But no need to do that in this PR, it's already quite large.

Reply:
that is a great idea!

Reply:
change made
commit 81a3787